File: src/main/scala/com/datastax/home_spotify/exercises/Exercise4.scala
package com.datastax.home_spotify.exercises

import com.datastax.home_spotify.exercises.Constants._
import com.datastax.home_spotify.exercises.Schema._
import com.datastax.spark.connector._
import com.datastax.spark.connector.rdd.CassandraRDD
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD

object Exercise4 extends BaseExercise {

  def main (args: Array[String]) {

    val sc = buildSparkContext(EXERCISE_4)


    /*
     * CREATE TABLE IF NOT EXISTS performers (
     *   name TEXT,
     *   ...
     *   country TEXT,
     *   ...,
     *   PRIMARY KEY(name)
     * );
     *
     *
     * CREATE TABLE IF NOT EXISTS albums (
     *   ...
     *   year INT,
     *   performer TEXT,
     *   ...,
     * );
     */

    val performers:RDD[(String,String)] = sc.cassandraTable(KEYSPACE, PERFORMERS)
      .select("name","country")
      .as((_:String,_:String))
      //TODO Filter out null countries or "Unknown" countries
      .filter(???)

    val albums:RDD[(String,Int)] = sc.cassandraTable(KEYSPACE, ALBUMS)
      .select("performer","year")
      .as((_:String,_:Int))
      //TODO Only takes years >= 1900
      .filter(???)


    // join performers with albums
    val join: RDD[(String, (String, Int))] = performers.join(albums)

    join
      //TODO RDD into a ((decade,country),1) using the computeDecade() pre-defined function
      .map[((String,String),Int)] {???}
      //TODO Reduce by key to count the number of occurrence for each key (decade,country)
      .reduceByKey(???)
      //TODO Flatten the tuple to (decade,country,count) triplet
      .map[(String,String,Int)] {???}
      .saveToCassandra(KEYSPACE, ALBUMS_BY_DECADE_AND_COUNTRY, SomeColumns("decade","country","album_count"))

    sc.stop()
  }

  def computeDecade(year:Int):String = {
    s"${(year/10)*10}-${((year/10)+1)*10}"
  }
}

File: src/main/java/fr/ippon/home_spotify/exercises/Exercise4.java
package fr.ippon.home_spotify.exercises;


import fr.ippon.home_spotify.entity.AlbumByDecadeAndCountry;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;


import static com.datastax.spark.connector.japi.CassandraJavaUtil.javaFunctions;
import static com.datastax.spark.connector.japi.CassandraJavaUtil.mapToRow;
import static fr.ippon.home_spotify.exercises.Schema.*;

import static fr.ippon.home_spotify.exercises.Constants.EXERCISE_4;

public class Exercise4 extends BaseExercise{

    public static void main(String[] args) {

        JavaSparkContext sc = buildSparkContext(EXERCISE_4);

        /*
         * CREATE TABLE IF NOT EXISTS performers (
         *   name TEXT,
         *   ...
         *   country TEXT,
         *   ...,
         *   PRIMARY KEY(name)
         * );
         *
         *
         * CREATE TABLE IF NOT EXISTS albums (
         *   ...
         *   year INT,
         *   performer TEXT,
         *   ...,
         * );
         */
        JavaPairRDD<String, String> performers = javaFunctions(sc)
                .cassandraTable(KEYSPACE, PERFORMERS)
                .select("name", "country")
                .filter(row -> {
                    String country = row.getString("country");
                    return country != null && !country.equals("Unknown");
                })
                .mapToPair(row -> new Tuple2(row.getString("name"), row.getString("country")));

        JavaPairRDD<String, Integer> albums = javaFunctions(sc)
                .cassandraTable(KEYSPACE, ALBUMS)
                .select("performer", "year")
                .filter(row -> row.getInt("year") >= 1900)
                .mapToPair(row -> new Tuple2(row.getString("performer"), row.getInt("year")));


        //Join performers with albums
        final JavaPairRDD<String, Tuple2<String, Integer>> joins = performers.join(albums);


        final JavaRDD<AlbumByDecadeAndCountry> result = joins
            // Map RDD into a ((decade,country),1) using the computeDecade() pre-defined function
            .mapToPair(join -> new Tuple2<>(new Tuple2<>(computeDecade(join._2()._2()), join._2()._1()), 1))
            // Reduce by key to count the number of occurrence for each key (decade,country)
            .reduceByKey((left, right) -> left+right)
            // Map the tuple into AlbumByDecadeAndCountry POJO
            .map(grouped -> new AlbumByDecadeAndCountry(grouped._1()._1(), grouped._1()._2(), grouped._2()));

        // Save back to Cassandra
        javaFunctions(result)
            .writerBuilder(KEYSPACE, ALBUMS_BY_DECADE_AND_COUNTRY, mapToRow(AlbumByDecadeAndCountry.class))
            .saveToCassandra();

        sc.stop();
    }

    private static String computeDecade(Integer year) {
        int lowerBound = (year/10)*10;
        int upperBound = ((year/10)+1)*10;
        return lowerBound+"-"+upperBound;
    }
}
